Adding MINC suport and PETPVC support

sulantha2006 opened the pull request:
With this, the minc-tools-1.9.15(latest stable), beast library and mni models are installed. 
Nipype MINC interface can be tested with this.

kaczmarj commented:
Hi @sulantha2006 - thank you for the pull request! Can you please merge the current master branch?

I see that MINC was compiled on relatively recent OS versions. Those binaries might not be compatible with older distros, which I would like to support as much as possible. It would be preferable to compile MINC on an older OS version and use those binaries for neurodocker. I am currently testing whether MINC compiled on CentOS 6 (glibc version 2.12) works on more recent Linux distributions (including debian-based distros). If it does, those binaries could be used in all or most cases, and users will not have to specify their Linux distribution. I am building MINC in a Docker image, following the specifications in [BIC-MNI/build_packages](https://github.com/BIC-MNI/build_packages), and I will submit a PR with my Dockerfile.

kaczmarj commented:
By the way, the travis tests will not pass right now because the secret environment variables necessary to log in to Docker are not available to branches outside of the original repo. I will fix that issue in `.travis.yml`.

sulantha2006 commented:
Do you need to add the docker repos in your hub account for the tests to complete?

kaczmarj commented:
@sulantha2006 - yes, you're right. I will modify the test-runner so that it builds the image from scratch if it cannot find it. I will let you know when I have made the changes.

I have been trying to compile MINC on CentOS 6 — those binaries *should* be compatible with most newer linux distributions — but I have not been successful. I will continue trying, and I might post an issue on MINC's github page. It would be ideal to not have to specify the distribution. Consider, for example, that a user might use a base image like continuumio/miniconda3, where it is not immediately clear what the linux distribution is (debian, in this case).

sulantha2006 commented:
I also added PETPVC as well.

kaczmarj commented:
@sulantha2006 - please merge current master. Test-runner should be fixed.

sulantha2006 commented:
Is it possible to create the docker repo for the PETPVC as well? I guess the tests fail because of this.

sulantha2006 commented:
Merged and the tests are passed.

kaczmarj commented:
@sulantha2006 - glad to see the tests are passing. in the future, i plan to expand the tests. right now, neurodocker mostly tests whether the software is installed and executable by a non-root user.

i am currently compiling minc-toolkit v2 on centos 6.

kaczmarj commented:
@sulantha2006 - I was able to build minc-toolkit-v2 version 1.9.15 on CentOS 6.9. I uploaded the [binaries to dropbox](https://www.dropbox.com/s/40hjzizaqi91373/minc-toolkit-1.9.15-20170529-CentOS_6.9-x86_64.tar.gz?dl=0). I would like to use these binaries for all Linux distributions. The shell script `minc/minc-toolkit-config.sh` should be sourced in the neurodocker entrypoint.

These libraries will have to be installed as well:

```
libGL.so.1
libICE.so.6
libSM.so.6
libX11.so.6
libXext.so.6
libXi.so.6
libXmu.so.6
libgomp.so.1
libjpeg.so.62
libopenjpeg.so.2
```

I found those libraries with this command, run in `/opt/minc/bin`:

```shell
find -type f -executable -exec /bin/bash -c "file -i '{}' | grep -q 'x-executable; charset=binary'" \; -print | xargs ldd | sed -n '/not\ found/p' | sed 's/=>.*//' | xargs -n1 | sort | uniq
```

Why are mincbeast and bci-mni models installed in addition to the minc-toolkit? It looks like they are both included in the toolkit.

kaczmarj commented:
I am rebuilding minc-toolkit-v2 now without openjpeg, so we won't have to install that as a dependency (it is not present in the debian stretch repositories).

sulantha2006 commented:
mincbeast is the command to run beast, but it needs the beast library
installed to function.
mni-models installs the 2009a/2009c versions of the MNI 152 template. I am
not sure this is also done when you make minc-toolkit.

S.

On Mon, Aug 28, 2017 at 12:10 PM, Jakub Kaczmarzyk <notifications@github.com
> wrote:

> I am rebuilding minc-toolkit-v2 now without openjpeg, so we won't have to
> install that as a dependency (it is not present in the debian stretch
> repositories).
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/kaczmarj/neurodocker/pull/79#issuecomment-325398518>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA-V976mStWCL1kGv47vnZA3qQ765_vlks5scuZcgaJpZM4O64uo>
> .
>

kaczmarj commented:
@sulantha2006 - I rebuilt the binaries without libopenjpeg, so we should be good to go. https://www.dropbox.com/s/40hjzizaqi91373/minc-toolkit-1.9.15-20170529-CentOS_6.9-x86_64.tar.gz?dl=0

Here are the missing shared libraries:

```
libGL.so.1
libICE.so.6
libSM.so.6
libX11.so.6
libXext.so.6
libXi.so.6
libXmu.so.6
libgomp.so.1
libjpeg.so.62

```

sulantha2006 commented:
The tests are passing. But I did not install the shared libraries. Should they be installed? If so, How should they be installed?

kaczmarj commented on neurodocker/interfaces/minc.py:
```diff
@@ -0,0 +1,128 @@
+        ent = _add_to_entrypoint("source /opt/minc/minc-toolkit-config.sh",
+                                 with_run=False)
+
+        cmd = self._get_install_cmd(minc_url, beast_url, models_09a_url, models_09c_url, ent)
```
Can you put `deps_cmd` and `cmd` in the same `RUN` instruction? I don't think it is necessary to have them as separate layers.

kaczmarj commented on neurodocker/interfaces/petpvc.py:
```diff
@@ -0,0 +1,92 @@
+                "\n&& mkdir /opt/petpvc"
+                "\n&& mv /tmp/PETPVC*/* /opt/petpvc"
+                "\n&& rm -r /tmp/PETPV* "
+               .format(petpvc_url=petpvc_url))
```
Correct me if I am wrong, but I think you can use the `--strip-components` option of `tar` here.

```shell
mkdir /opt/petpvc
curl ... | tar xz --strip-components=1 -C /opt/petpvc
```

kaczmarj commented on neurodocker/neurodocker.py:
```diff
@@ -132,6 +132,11 @@ def list_of_kv(kv):
             " keys are version and matlab_version."),
+        "minc": (
+            "Install MINC. Valid keys is version (required). Only the 1.9.15"
+            " version is supported at this time."),
```
Can you change this to `Only version 1.9.15 is supported at this time.`

kaczmarj commented:
Regarding the shared libraries, search the `apt` and `yum` repositories for the packages that install those libraries.

some examples (untested):
- libGL.so.1
  - apt-get: `libgl1-mesa-dev`
  - yum: `mesa-libGL-devel`
- libICE.so.6
  - apt-get: `libice6`
  - yum: `libICE`

and so on. Let me know if you need help with this.

sulantha2006 commented on neurodocker/interfaces/minc.py:
```diff
@@ -0,0 +1,128 @@
+        "1.9.15": "https://www.dropbox.com/s/40hjzizaqi91373/minc-toolkit-1.9.15-20170529-CentOS_6.9-x86_64.tar.gz?dl=0",
+    }
+    BEAST_URL = {
+        "latest": "http://packages.bic.mni.mcgill.ca/tgz/beast-library-1.1.tar.gz",
```
I am not sure the beast_library is installed with the minc-toolkit. I know it installes the beast binaries, but not the library. And without the libraries, beast command cannot function.

kaczmarj replied:
You are right about the beast library. Sorry about that.

kaczmarj replied:
The url for minc-toolkit should be changed to https://dl.dropbox.com/s/40hjzizaqi91373/minc-toolkit-1.9.15-20170529-CentOS_6.9-x86_64.tar.gz

kaczmarj replied:
`dl` replaces `www`

sulantha2006 commented on neurodocker/interfaces/minc.py:
```diff
@@ -0,0 +1,128 @@
+        "latest": "http://packages.bic.mni.mcgill.ca/tgz/beast-library-1.1.tar.gz",
+    }
+    MODELS_09a_URL = {
+        "latest": "http://www.bic.mni.mcgill.ca/~vfonov/icbm/2009/mni_icbm152_nlin_sym_09a_minc2.zip",
```
Same here. I could not find the mni models in the tar.gz you created. Am I missing it?

kaczmarj replied:
You are also right about the models. Sorry about that.

sulantha2006 commented:
I added the shared libs, just kept the imagemagik there as I am not sure which minc command will it break. May be we can remove it later. Also, the tar strip components was not working, I reverted it back for now. 
Tests are passing.

kaczmarj commented:
I tested the shared libs on a debian stretch docker image.

These should be enough for `apt`:
- `libgl1-mesa-dev libice6 libsm6 libx11-6 libxext6 libxi6 libxmu6 libgomp1 libjpeg62`

sulantha2006 commented:
Were you able to check if installing only this shared libs breaks any minc commands? I know the test will pass. But I am not sure if any of the other minc commands will fail.

kaczmarj commented:
I ran a command to check if any shared libraries were missing, and all of the minc binaries found their required shared libs. Here is the command:

```shell
find /opt/minc -type f -executable -exec /bin/bash -c "file -i '{}' | grep -q 'x-executable; charset=binary'" \; -print | xargs ldd | sed -n '/not\ found/p' | sed 's/=>.*//' | xargs -n1 | sort | uniq
```

kaczmarj commented:
For yum, it looks like these libraries will suffice:

- `mesa-libGL-devel libICE libSM libX11 libXext libXi libXmu libgomp libjpeg-turbo`

I replaced `libgomp1` with `libgomp`, because for some reason, libgomp.so.1 was not being found after installing `libgomp1` (odd). Using the other one fixed that.

kaczmarj commented:
Regarding `petpvc`, here is an example that uses `--strip-components`:

```Dockerfile
RUN echo "Downloading PETPVC..." \
    && mkdir /opt/petpvc \
    && curl --retry 5 -sSL https://github.com/UCL/PETPVC/releases/download/v1.2.0-b/PETPVC-1.2.0-b-Linux.tar.gz \
    | tar zx --strip-components=1 -C /opt/petpvc
ENV PATH=/opt/petpvc/bin:$PATH
```

sulantha2006 commented:
Is it possible to rebuild MINC without ANTS? . By default, MINC-toolkit includes an older version of ANTS and builds it too. So, the ANT commands in /opt/ants get masked by the one ones in /opt/minc/bin. 
Can you rebuild minc-toolkit wihtout ANTS and let me know the dropbox link. 

Thanks.

kaczmarj commented:
Sure, I will rebuild it tonight.

kaczmarj commented:
One thing we can try is modify the file `minc-toolkit-config.sh` to append MINC's path to the end of `$PATH` instead of to the beginning. Then, other C3D and ANTs installations should be found first. I am not sure whether this will cause issues for MINC, but we can test this.

sulantha2006 commented:
Yeah. Let's go ahead with the minc-toolkit-config.sh option. Can you modify it and rebuild the tar. ?

kaczmarj commented:
@sulantha2006 - I modified the shell and cshell files. The link to the binaries should be the same:
https://dl.dropbox.com/s/40hjzizaqi91373/minc-toolkit-1.9.15-20170529-CentOS_6.9-x86_64.tar.gz

Can you test this with real data?

sulantha2006 commented:
I tested this. The minc tools do not shadow the ANTS command anymore. Also, I tested a couple of minc commands with real data, they work as expected.

kaczmarj commented:
@sulantha2006 - thank you very much for the pull request. Looks good to me.
